File: crypto/benchs/bench-bitperm.py
import timeit
import itertools
import cProfile

import sys, os.path
root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
tools_dir = os.path.join(root_dir, 'src/tools')
sys.path.append(tools_dir)

from permgen import * 
from permpy import *

def bench_nbit_perm(n, masks_n, samples_per_mask):

    # generate permutations to execute
    raw_masks = [(x, s, d) for x, s, d in generate_random_masks(n, masks_n)]

    # generate samples numbers to that will be fed into permutations
    total_samples = masks_n * samples_per_mask
    samples = [random.randint(0, 2**n - 1) for _ in range(total_samples)]
    samples_data = (ctypes.c_uint64 * total_samples)(*samples)

    print("Benchmarking Bitperms...")

    #########################
    # Benchmark Dumb Python #
    #########################

    t = time.time()
    i = 0
    for _, src, dst in raw_masks:
        for _ in range(samples_per_mask): 
            dumb_perm(samples[i], src, dst)
            i += 1
    duration = time.time() - t
    python_speed = duration / (masks_n * samples_per_mask)

    print("\nPython Results:")
    print("---------------")
    print("Total seconds:\t", duration)
    print("Seconds / Call:\t", python_speed)
    print("Sanity Check:\t", hex(dumb_perm(samples[-1], raw_masks[-1][1], raw_masks[-1][2])))
    print('')

    #########################
    # Benchmark Dumb Native #
    #########################

    srcs = [[ord(a) - ord('A') for a in s] for x, s, d in raw_masks]
    dsts = [[ord(a) - ord('A') for a in d] for x, s, d in raw_masks]
    srcs_data = [(ctypes.c_uint8 * len(x))(*x) for x in srcs]
    dsts_data = [(ctypes.c_uint8 * len(x))(*x) for x in dsts]
    srcs_ptrs = (ctypes.POINTER(ctypes.c_uint8) * masks_n)(*srcs_data)
    dsts_ptrs = (ctypes.POINTER(ctypes.c_uint8) * masks_n)(*dsts_data)
    native_dumb_speed = bench_perm_bitstring_dumb(
            n, srcs_ptrs, dsts_ptrs, masks_n, samples_data, samples_per_mask)

    #########################
    # Benchmark Fast Native #
    #########################

    # convert the list of lists to a list of c, uint64 arrays
    mask_data = [(ctypes.c_uint64 * len(x))(*x) for x, y, z in raw_masks]

    # create an array of Uint64 pointers from the data arrays
    mask_ptrs = (ctypes.POINTER(ctypes.c_uint64) * masks_n)(*mask_data)
    native_fast_speed = bench_perm_bitstring(
            n, mask_ptrs, masks_n, samples_data, samples_per_mask)

    time.sleep(1)

    # Print final results
    print("\nSpeed-up Factor Results")
    print("-----------------------")
    print("Dumb C vs Dumb Python\t", round(python_speed / native_dumb_speed, 2))
    print("Fast C vs Dumb C\t", round(native_dumb_speed / native_fast_speed, 2))
    print("Fast C vs Dumb Python\t", round(python_speed / native_fast_speed, 2))

if __name__ == "__main__":
    bench_nbit_perm(64, 1000, 10000)

File: crypto/benchs/bench_bitperm.c
#include "bitperm.h"

uint64_t
perm_bitstring_dumb(uint64_t x, uint8_t n, uint8_t* src, uint8_t* dst)
{
    uint64_t res = 0; 
    uint8_t* bits = malloc(n * sizeof(uint8_t));
    uint8_t* permbits = malloc(n * sizeof(uint8_t));
    uint8_t bit_i = n - 1;  
    uint64_t i = 1;
    uint8_t j; 
    uint64_t max = n == 64? 0 : 1ULL << n; 

    /* Set the bit in the array if its set */
    for (i = 1; i != max; i <<= 1, bit_i--) 
        bits[bit_i] = (i & x)? '1' : '0';               

     /* For each source bit, find the corresponding destination bit
      * and move the src's value into the desintations' index. 
      */ 
    for (i = 0; i < n; i++) 
       for (j = 0; j < n; j++) 
            if (dst[j] == src[i]) 
                permbits[j] = bits[i];              

    /* Set the bit if it's set in the array */
    bit_i = n - 1;
    for (i = 1; i != max; i <<= 1, bit_i--) 
        if (permbits[bit_i] == '1') res = res | i;     

    return res; 
}

/**
 * bench_perm_bitstring_dumb - benchmark dumb bitstring permutations 
 *
 * @n: The number of bits which will be permuatated.
 * @srcs: Nested array of source permutation. E.g. [[1, 2, 3], [2, 3, 1], ...]]
 * @dsts: Nested array of dest permutation.   E.g. [[3, 2, 1], [2, 1, 3], ...]]
 * @perms_n: Length of srcs and dsts arrays.
 * @samples: Pointer to an array of values to permute.
 * @samples_per_perm: Number of samples to run for each permutation mask
 *
 * Note: samples needs to point to a block of uint64_t's of size 
 *       samples_per_perm * perms_n = total number of samples
 */
double
bench_perm_bitstring_dumb(uint8_t n, uint8_t** srcs, uint8_t** dsts, int perms_n,
        uint64_t* samples, int samples_per_perm)
{
    /* Start timer */
    uint64_t start_time = __rdtsc();

    /* Do computations */
    int i, j;
    uint64_t x;
    for (i = 0; i < perms_n; i++, srcs++, dsts++) 
        for (j = 0; j < samples_per_perm; j++, samples++) 
            x = perm_bitstring_dumb(*samples, n, *srcs, *dsts);

    /* Print Results */
    uint64_t elapsed = (__rdtsc() - start_time);
    double freq = 3600.033e6;
    double seconds = (elapsed / freq) ;
    double seconds_per_call = seconds / (perms_n* samples_per_perm);

    printf("Dumb C Results:\n");
    printf("-----------------\n");
    printf("Ticks elapsed: \t%ld\n", elapsed);
    printf("Total seconds: \t%.17g\nSeconds / Call: %.17g\n", seconds, seconds_per_call);
    printf("Sanity Check: \t0x%lx\n\n", x); 

    return seconds_per_call;
}

/**
 * bench_perm_bitstring - benchmark bitstring permutations given a list of masks.
 * @n: The number of bits which will be permuatated
 * @mask: A list of lists of masks
 * @mask_n: Length of mask list
 * @samples: The number of randomly generated samples which will be run for each mask.
 */
double
bench_perm_bitstring(uint8_t n, uint64_t** masks, int masks_n,
        uint64_t* samples, int samples_per_mask)
{
    /* Start timer */
    uint64_t start_time = __rdtsc();

    /* Do computations */
    int i, j;
    uint64_t x;
    for (i = 0; i < masks_n; i++, masks++)
        for (j = 0; j < samples_per_mask; j++, samples++)
            x = perm_bitstring(*samples, n, *masks);

    /* Print Results */
    uint64_t elapsed = (__rdtsc() - start_time);
    double freq = 3600.033e6;
    double seconds = (elapsed / freq) ;
    double seconds_per_call = seconds / (masks_n * samples_per_mask);

    printf("Fast C Results:\n");
    printf("-----------------\n");
    printf("Ticks elapsed: \t%ld\n", elapsed);
    printf("Total seconds: \t%.17g\nSeconds / Call: %.17g\n", seconds, seconds_per_call);
    printf("Sanity Check: \t0x%lx\n\n", x);

    return seconds_per_call;
}
